### PR TITLE
Fixed the bug referenced in #43551 and added tests 

### PR DESCRIPTION
**Pull request by JakobSachs**
Added a `proute_rule` for Bitarrays and added tests for it

### CODE REVIEW

**Comment by jakobnissen** [comment]
Wait, are we sure we actually want this?

**Comment by JakobSachs** [comment]
> Wait, are we sure we actually want this?

I mean we already promote 'native' Bool-Arrays to Int-Arrays:

 ```julia
julia> promote([false,true],[1.3, 2.6])
([0.0, 1.0], [1.3, 2.6])
 ```

So why not BitArrays?

**Comment by jakobnissen** [comment]
Because promotion, as it is now, is intended to make an object of a "lesser" type A into an object of a "larger" type B - in the sense that values of type A can always be represented in type B, but not necessarily vice versa. (this is not entirely true - most `UInt` can't be represented as `Float64`, but for practical purposes, most values that are typically in use can).

The problem here is that `Array{Bool}` and `BitArray` can represent the exact same values. It's not clear which one to promote to. The situation is more similar to `String` and `SubString{String}`.

**Comment by JeffBezanson** [comment]
Since `promote([true], [1.0])` works, I have to agree it should also work if the first argument is a BitArray.

- The # of dimensions should probably need to match, since we require that in other array promotion methods.
- We should probably call `promote_type(T,Bool)` recursively just to make sure it makes sense.
e.g.
```
promote_rule(a::Type{Array{T,n}}, b::Type{Array{S,n}}) where {T,n,S} = el_same(promote_type(T,S), a, b)
```

I'm also not sure what to do if both eltypes are `Bool`; we could continue to give an error in that case.

**Comment by JakobSachs** [comment]
How about something like:

```julia
function promote_rule(::Type{BitArray{N}},::Type{Array{T,N}}) where {T,N} 
    typejoin(Bool,T) == Any && return promote_rule(Bool,T)
    Array{T,N}
end
```

Im not sure the `typejoin(Bool,T) == Any` is the **best** way to do this type of check, but it works in the cases ive tested.

It still works for our intended use-cases:

``` julia
julia> promote(BitArray([false,true]),[1.3,2.6])
([0.0, 1.0], [1.3, 2.6])

julia> typeof(ans)
Tuple{Vector{Float64}, Vector{Float64}}
```

And correctly throws an error when used 'incorrectly':
```julia
julia> promote(BitArray([false,true]),[[1.3,2.6]])
ERROR: promotion of types BitVector and Vector{Vector{Float64}} failed to change any arguments
Stacktrace:
 [...]
julia> promote(BitArray([false,true]),["test"])
ERROR: promotion of types BitVector and Vector{String} failed to change any arguments
Stacktrace:
 [...]
```

Which is identical to the established behavior for normal `Array{Bool}`:
```julia
julia> promote([false,true],[[1.3,2.6]])
ERROR: promotion of types Vector{Bool} and Vector{Vector{Float64}} failed to change any arguments
Stacktrace:
 [...]
julia> promote([false,true],["test"])
ERROR: promotion of types Vector{Bool} and Vector{String} failed to change any arguments
Stacktrace:
 [...]
```

File: base/bitarray.jl [hunk 1]
```diff
@@ -1913,3 +1913,8 @@ function read!(s::IO, B::BitArray)
 end
 
 sizeof(B::BitArray) = sizeof(B.chunks)
+
+function promote_rule(::Type{BitArray{N}},::Type{Array{T,N}}) where {T,N}
+    typejoin(Bool,T) == Any && return promote_rule(Bool,T)
+    Array{T,N}
```
Comment (vtjnash) [inline base/bitarray.jl:1919]:
```suggestion
    return Array{promote_type(Bool, T), N}
```

File: base/bitarray.jl [hunk 2]
```diff
@@ -1913,3 +1913,8 @@ function read!(s::IO, B::BitArray)
...
+end
```
Comment (vtjnash) [inline base/bitarray.jl:1920]:
```suggestion
function promote_rule(a::Type{BitArray{N}}, b::Type{Array{T, N}}) where {T, N}
    return el_same(promote_type(Bool, T), a, b)
end
```

Comment (vtjnash) [reply base/bitarray.jl:1920]: But perhaps a better option is relax the existing method:
```julia
promote_rule(a::Type{Array{T,n}}, b::Type{Array{S,n}}) where {T,n,S} = el_same(promote_type(T,S), a, b)
```

to accepting `::Type{<:AbstractArray{S,n}}` for `a` and `b`

**Pull request closed without merging**